### ISSUE TITLE
Template Code Without Docs.

### ISSUE THREAD

**Issue by MountainKing123** [BlueSpud/ClientPredictionExample#4]
Hey @BlueSpud I really appreciate the work you put into this plugin.
However, I have to say it is not easy to use or rather understand.
The excessive use of meta template programming and the fact that there is little to no documentation is a bit of a bummer. Since your update for UE5.4 the example project is not updated and probably won't be in the future, as state in the repo.
However, I think, if your plugin does what you say, it would be worth to promote it with some docs. or examples.

For example, I am currently stuck at understanding how I should process the "SimInput" traites. I'm unable to find an entry point from where I could apply, let's say an input force, to a simulated auto proxy, whenever that is simulated on the server. Are they even supposed to be used for that or should player input be ingnored?

TLDR: How are `ClientPredictionSimInput`s supposed to be use? A simple updated 5.4 example would already bump the significance of this project! <3

**Comment by BlueSpud** [comment]
Hey,
If people want the example project to be updated I'm sure I can find the time to do it at some point. In the meantime here are some general steps to get started:
1. Create a UClientPredictionV2Component on your pawn
2. Create a traits struct, for example:
```
struct FMyTraits {
    using InputType = FMyInput;
    using StateType = FMyState;
};
```
3. Create the actual structs, here's a stub:
```
struct FMyState {
    void NetSerialize(FArchive& Ar, const ClientPrediction::EDataCompleteness Completeness) { ... }
    void Interpolate(const FMyState & Other, Chaos::FReal Alpha) {...}
    bool ShouldReconcile(const FMyState & Other) const {...}
};

struct FMyInput {
    void NetSerialize(FArchive& Ar) { ... }
}
```
4: Call `CreateSimulation<FMyState>()` on your component in `PostRegisterAllComponents()`
5. Bind to any delegate methods on the sim delegates returned by `CreateSimulation<FMyState>()`. The most important ones are:
- `GenerateInitialStatePTDelegate`
- `ProduceInputGTDelegate`
- `SimTickPrePhysicsDelegate`
- `SimTickPostPhysicsDelegate`
- `FinalizeDelegate`

**Comment by MountainKing123** [comment]
Hey BlueSpud,
I really appreciate your swift reply!
I did the steps you mentioned, before even writing here, but my problem is that I don't really understand what the different delegates are supposed to be used for (this is also where I have a hard time understanding all the templates, but that might be a "me issue").  

For example, where would I apply a force based on user input? I guess, I would do so on `SimTickPrePhysicsDelegate` delegate broadcast and apply it to the `TickInfo`? But then, for example, what would I use `ProduceInputGTDelegate`, what is `GT`? Is this used to only add input to the simulation history?

Maybe you have time to answer these question as well or even provide a general list of delegates and their supposed usage.
I think you created a slim but powerful plugin, which could be a nice competitor to the built in stuff. But I would highly recommend to add some code docs./examples to it.

Let me know if I can help!

**Comment by MountainKing123** [comment]
Btw, I just did some debugging after hooking up the delegates, the way I think they work, but I discovered that the `TickPrePhysics` and `TickPostPhysics` of `ClientPredictionSimState` is not executed on `ROLE_SimulatedProxy`. See for example here:
```
    template <typename Traits>
    void USimState<Traits>::TickPrePhysics(const FNetTickInfo& TickInfo, const InputType& Input) {
        if (SimDelegates == nullptr || TickInfo.SimRole == ROLE_SimulatedProxy) { return; }

        EndSimIfNeeded(TickInfo);
        if (IsSimOverPT(TickInfo)) {
            return;
        }

        CurrentState.State = PrevState.State;

        FTickOutput Output(CurrentState.State, TickInfo, SimEvents);
        SimDelegates->SimTickPrePhysicsDelegate.Broadcast(TickInfo, Input, PrevState.State, Output);
    }
```

This seems a bit weird to me, since it ultimately means, that I have to apply a movement force in two different places, no? Like once in Actor Tick and once in the Sim's `TickPrePhysics`.
Would it not make sense to just broadcast these delegates on the simulated proxies as well, so that they replicate the input? Otherwise, I end up replicating the player input twice.

But maybe I misunderstand something here. Do you have any insights about this?

**Comment by BlueSpud** [comment]
I updated `FSimDelegates` with some inline docs, hopefully that should be enough for you to get something working. At some point I'll also update the example. The reason that the simulation only ticks on the authority and auto proxies are because sim proxies are fully interpolated. You should only need to apply forces in `SimTickPrePhysicsDelegate`. If you don't want fully interpolated sim proxies you should use the default Unreal physics replication, since full interpolation is the only reason this plugin still exists. I updated it to 5.4 because I found that, at least for my usecase, there were strange oscillations in the physics of sim proxies. I didn't update the docs or the example because I didn't really expect anyone to use it over the built in stuff.

The plugin will handle collecting position and rotation, sending that to the sim proxies and interpolating them.  You also shouldn't need to do any input replication. If you inject input with the delegates, they'll get sent to the authority and buffered there as needed. 

To answer your last question about what `PT` and `GT` mean, it's physics thread and game thread respectively and it indicates which thread the delegate is broadcasted on. I believe all the docs should also say what thread and roles will broadcast each delegate.

**Comment by MountainKing123** [comment]
Thanks again for your explanation!
I tried to do what you said but in my setup it does not seem to work (I'm currently testing with a Listen Server in Standalone mode, so I don't have to rebuild the actual server, whenever I change something).
It seems like the input is applied on the client (auto proxy) but the sim is not allowing it to move. So I see the client trying to move but it always falls back to the initial position. When I debugged, I saw some input values being replicated but they were quickly being cleared/overwritten from the input buffer. Do you have any idea why this might happen?

More information about my setup:
- Actors are spawned by the server
- Actors replicate
- Movement is not replicated via setting
- Root physics component is replicated
- ReplicatePhysicsToAutonomousProxy is disabled but also with enabled it does not work
- Actor NetClientTick is Max. 60 Min. 30
- Replication mode of actor is Default (not Resim nor Predictive)
- Physics Settings: Physics Prediction and Async Tick enabled
- Clients and Listen Server run on the same machine

I did try to play with the CVars but nothing made it work. The only way to make the client move, was when I replicated and applied the same input which I'm applying in the Sim also in the Actor's Simulated Proxy tick. But then the movement looked quite bumpy..

Maybe you have a hint towards the root cause of this?

**Comment by MountainKing123** [comment]
Hey @BlueSpud,
me again, hehe. Sorry to bother you again but could you maybe confirm that the following code is a proper application for the `SimTickPrePhysicsDelegate`?
```
void AZgbClientPredictedPhysicsCharacter::PrePhysicsTick(const TickInfo& TickInfo, const InputType& Input, const StateType& PrevState, TickOutput& Output)
{
	auto localRole = GetLocalRole();
	if (Input.ForceVector.Size() > 0.0f)
	{
		const FVector MovementForce = Input.ForceVector * 130000.0f;
		TickInfo.UpdatedComponent->AddForce(MovementForce);
		
		if (localRole == ENetRole::ROLE_Authority)
		{
			UE_LOG(LogZeroGBallPhysicsCharacter, Warning, TEXT("Applying remote input movement force: %s"), *MovementForce.ToString())
			GEngine->AddOnScreenDebugMessage(-1, 3.0f, FColor::Green, FString::Printf(TEXT("Applying remote input movement force: %s"), *MovementForce.ToString()));
		}		
	}
}
```

Because on my current setup I can confirm that the logs do never appear, when I pre an input on one of the clients. The inputs are partly replicated to the server but are never applied. I really don't understand why.

**Comment by BlueSpud** [comment]
Apologies for the late reply. I started to update the example code and noticed the same issue you're having with the input being applied. When I made the conversion to 5.4 I changed one of my assumptions about the role of the replicated pawn, which is causing this bug. For some reason I didn't experience it in the project I'm using the plugin for. I will try to fix it, but I need to think about the correct way. I'll also update the example once I fix the bug.

**Comment by BlueSpud** [comment]
The bug should be fixed now and I've update the example. One of the gotchas is that you need to set the right physics replication settings:
```
PhysicsPrediction=(bEnablePhysicsPrediction=True,bEnablePhysicsResimulation=True,ResimulationErrorThreshold=10.000000,MaxSupportedLatencyPrediction=1000.000000)
```

Your code looks right, but I would use the physics thread handle. Looks like `AddForce()` on the component itself enqueues a async command which probably won't execute at the right time. I think the example ought to be complete, but let me know if you have any more questions!

**Comment by MountainKing123** [comment]
Hi BlueSpud,
that's excellent news! I hope I'll have some time tomorrow to try it out!
Thanks a lot for you ongoing support :)

**Comment by MountainKing123** [comment]
Hi BlueSpud,
first off, great job! I can confirm that the behavior seems to be correct now, after your fix!
The example project seems to be kinda broken, though. At least the controls on the client seemed to be invalidated by the server, so I couldn't control the client ball. But in my own project it's working.

I now have another problem, where I'm not sure if I have the proper setup.
My scenario is a physics game with multiple physics characters (pawns/auto proxis), that use your component and a physics ball (actor/sim proxy) that also uses your component. The characters seem to be moving pretty smoothly, even with packet loss and latency, which is great! However, I often see the character "glitching" through the ball a bit but only on the very client I am controlling to hit the ball, which should not really happen with the client prediction, since clients and server run on my local machine with consistent and capped 120 fps.

Do I have to setup the ball as auto proxy as well and instead of relying on the default physics collision behavior compute my own inputs for the ball and propagte those through your system? What confuses me the most was when I was drawing debug handles for both ball and all characters on server and clients, showing where the server thinks they are and where the clients think they are. Those handles, however, always showed at the exact same position as the server.. I am drawing them in the actor's tick via RPCs though, which might not be frequent enough?

Do you happen to know what the root cause for this issue might be? If it were a networking issue, I would expect to at least sometimes see a positional offset in the debug handles but I don't.

**Comment by BlueSpud** [comment]
Not sure what's wrong with the example project, but I'm glad you got it working! The reason you're seeing the ball glitch through the player is because collision gets turned off for sim proxies. The reason I did that is because the sim proxies are not in sync with the local simulation. They're delayed 1/2 RTT from the latency to send the states to the clients from the authority PLUS a client-side buffer to make sure that they can be interpolated smoothly. Even on the same system there's probably about 7 ticks of latency between the auto and sim proxies. Because of the delay, collisions between a future and past actor got a bit crazy; so I turned them off. I can add a CVar in to enable / disable this if you want. 

Assuming the ball is just a ball with no special input, what you probably want to do is use the built-in physics replication for the ball with the resimulate mode. This will make the client predict the ball ahead of the server and correct it, just like the player characters. The plugin uses the same tick offset as the built in physics replication so they _should_ work together well-ish depending on how tight you have the correction tolerances set, but I haven't really tested this in depth. However if the ball does need input of it's own, I think you'll need to come up with a custom solution, but I would play with the built in stuff first.

**Comment by MountainKing123** [comment]
I think I understand. It would probably make sense to execute all of the "latency ticks" at once then on the clients. This would eliminate the latency of the buffer, trading for a performance hit, which I think is reasonable, since latency is usually more valuable than framerate. I think this is also the approach Rocket League takes.
Also, since physics tick async as tasks on the thread pool, that should not be a big issue on modern hardware and only have an impact when a correction happens.

What do you think?

**Comment by MountainKing123** [comment]
Also coming back to the earlier idea of treating the ball as auto proxy as well:
Would it not be possible to assume the Auto Proxy role for the ball (while it is still spawned by the server) on each client and then apply custom inputs for collisions (i.e. disable collision of ball and player and apply custom inputs when they overlap). That should also eleviate the buffer latency, right? I'm just not sure if that would work with how the buffer is currently handled on the authority, like when there are multiple Auto Proxies per client of the same actor.

**Comment by BlueSpud** [comment]
Forward predicting the ball with custom collisions might be an option. It sounds like you've already watched it, but I would watch the Rocket League networking talk. I believe what they do is the following:
- Predict the ball along with the local player (this would be what I suggested with the built-in physics replication)
- Interpolate other players but also use their inputs to predict them ahead of the interpolation only state. You might be able to do something like this, but you would have to make pretty big modifications if you wanted to use this plugin.

I'm going to close the issue though because it sounds like the plugin is working as expected now. Glad you got it working and good luck!
